src/main/java/net/sandrohc/schematic4j/builder/SchematicBuilder.java

```diff
@@ -0,0 +1,252 @@
+package net.sandrohc.schematic4j.builder;
+
+import net.sandrohc.schematic4j.SchematicFormat;
+import net.sandrohc.schematic4j.exception.NoBuilderFoundException;
+import net.sandrohc.schematic4j.nbt.io.NamedTag;
+import net.sandrohc.schematic4j.schematic.Schematic;
+import net.sandrohc.schematic4j.schematic.types.*;
+
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+/**
+ * A utility class for building schematics.
+ */
+public abstract class SchematicBuilder {
+
+    protected SchematicBuilder() {}
+
+    protected SchematicFormat format;
+
+    protected String name;
+    protected String author;
+    protected int width;
+    protected int height;
+    protected int length;
+    protected SchematicBlockPos offset = SchematicBlockPos.ZERO;
+    protected SchematicBlock[][][] blocks = new SchematicBlock[][][]{};
+    protected List<SchematicBlockEntity> blockEntities = new ArrayList<>();
+    protected List<SchematicEntity> entities = new ArrayList<>();
+
+    /**
+     * Initializes a builder for a schematic. You can use this function to generate a schematic file.
+     * @param format The Schematic format.
+     * @return The builder.
+     * @throws NoBuilderFoundException If no builder was found for the desired schematic format.
+     */
+    public static SchematicBuilder create(SchematicFormat format) throws NoBuilderFoundException {
+        SchematicBuilder builder = format.createBuilder();
+        builder.format = format;
+        return builder;
+    }
+
+    /**
+     * Sets the name of the schematic.
+     */
+    public <R extends SchematicBuilder> R name(String name) {
+        this.name = name;
+        return (R) this;
+    }
+
+    /**
+     * Sets the author of the schematic.
+     */
+    public <R extends SchematicBuilder> R author(String author) {
+        this.author = author;
+        return (R) this;
+    }
+
+    /**
+     * Sets the dimensions of the schematic.
+     * @param width Schematic width (X axis)
+     * @param height Schematic height (Y axis)
+     * @param length Schematic length (Z axis)
+     */
+    public final <R extends SchematicBuilder> R dimensions(int width, int height, int length) {
```

Comment:
The dimensions and the block setter will clash with each other. I can think of three approaches:
1. Add checks to each function to make sure dimensions are set before blocks. Would work fine but it's not very user friendly, as one doesn't expect a builder setter to be fallible
2. Force dimensions to be specified on builder creation. Creation would be a bit clunkier, but everything else would be simpler
3. Make it so the blocks are not dependent on knowing the dimensions, and either check or ignore blocks outside of bounds when building the schematic. Feels like we're just moving the check from the setter methods to the builder method

I'd like to hear your opinion.

Reply:
Tricky one, but after thinking about it for a bit, I think people using the API will be self-aware of setting a block out of bounds, so I don't think a direct check is necessary.

I suggest an intermediary build function which will perform a few checks (also checks unrelated to blocks). This function should check all blocks to see if any are out of bounds, and if they are, those blocks should be ignored in the result and a simple non-halting warning should be logged. What do you think?

Reply:
I gave this problem a bit more thought. Most consumers of this library will not know or care what the schematic dimensions are. Blocks should always take precedence over any dimensions set. This means that we should still respect the dimensions when set, except when there's a block outside of bounds. In such cases, the dimensions should be expanded accordingly. With this approach, we maintain valid dimensions at all times and we never discard any blocks.

For this to work, we need three things:
- Store block data in a dimension-agnostic way. This means storing the blocks and coordinates in a list (or something else equally appropriate) instead of the 3D array it is currently using. The 3D array will force us to resize it on every dimension change, which will become very common after the changes I'm proposing
- Check for coordinates outside of bounds on block insertion, and update the dimensions accordingly
- Check (and fail with an exception) for negative coordinates on block insertion. Library consumers should set an offset and normalize the coords they send us instead

Reply:
Great ideas. Working on its implementation.

With this in mind, do you think we could get rid of the builder dimension setters? With this system in place, the schematic would automatically adapt to the most compact dimensions possible. Would further simplify things for library consumers as they don't have to think about dimensions at all.

Reply:
Should be fine to remove the dimension setters. For the ones that need more control over the dimensions, they can still modify them in the generated schematic.

Reply:
👍 I've already finished coding it, I'll remove the dimension setters and push it later when I get home

Reply:
Done.